### PR TITLE
Plasma: Abstract CommitmentData to Interface

### PR DESCRIPTION
**Pull request by axelKingsley**
# What
Abstracts Keccak and Generic Commitment types to a `CommitmentData`, and wires it throughout the project

# Why
`op-node` must support different modes of operation for Generic and Keccak commitments, which are signaled by a type byte at the front of the data.

# How
Now, in all places where a commitment is decoded, the package function `DecodeCommitmentData` is used. This handles the type check, does some basic validation, and then calls down to the implementation-specific decoder.

And, in all places where a commitment is *encoded*, `NewCommitmentData` is used. This function takes the `CommitmentType` to decide which concrete type to use.

The `CommitmentData` interface replaces all reference to concrete types in existing Plasma/DA behaviors.

Outside of these activities, all access to the `CommitmentData` happens through interface-defined functions which Keccak and Generic already supported -- `Encode, Verify, TxData`

# Omitting
Not featured in this PR is any feature-flagging from the node for the expected Commitment Type. @trianglesphere has a plan in mind to group the config details, so that will be wired in then. The DA Client similarly is set to only produce Keccak commitments.

# Testing
Tests were written for the CommitmentData interface to confirm it creates the right type of commitment and handles errors as expected.

DA server tests are fixed to type-assert Keccak type Commitments, as the server components are designed to only use those. When flags are wired into the `Next` and `SetInput` functions, the test can be multiplied by both Commitment Types. The Mock DA Server is already set to handle that with this PR, you can just overload the Commitment Type the mock uses.

### CODE REVIEW

File: op-plasma/commitment.go [hunk 1]
```diff
@@ -16,12 +16,76 @@ var ErrCommitmentMismatch = errors.New("commitment mismatch")
...
+		return DecodeGenericCommitment(data)
+	default:
+		return nil, ErrInvalidCommitment
+	}
+}
+
+// NewKeccak256Commitment creates a new commitment from the given input.
+func NewKeccak256Commitment(input []byte) Keccak256Commitment {
+	return Keccak256Commitment(crypto.Keccak256(input))
+}
```
Comment (axelKingsley) [inline op-plasma/commitment.go:75]: I renamed `Keccak256` to `NewKeccak256Commitment` to stay consistent with other factory functions.

File: op-plasma/commitment.go [hunk 2]
```diff
@@ -45,31 +104,17 @@ func (c Keccak256Commitment) Verify(input []byte) error {
...
-	}
-	return c, nil
-}
-
-// GenericCommitment is the default commitment type for op-plasma.
-type GenericCommitment []byte
-
-func NewGenericCommitment(commitment []byte) GenericCommitment {
-	return commitment
+	return commitment[:], nil
```
Comment (axelKingsley) [inline op-plasma/commitment.go:123]: This function moved upward slightly for more consistent ordering (New and Decode before struct functions), and the commitment data no longer strips the first byte because that is the responsibility of the Interface.

File: op-plasma/commitment.go [hunk 3]
```diff
@@ -24,9 +24,68 @@ const (
...
+type CommitmentData interface {
+	Encode() []byte
+	TxData() []byte
+	Verify(input []byte) error
+}
+
+// Keccak256Commitment is an implementation of CommitmentData that uses Keccak256 as the commitment function.
 type Keccak256Commitment []byte
 
+// Keccak256Commitment is an implementation of CommitmentData that treats the commitment as an opaque bytestring.
```
Comment (trianglesphere) [inline op-plasma/commitment.go]:
```suggestion
// GenericCommitment is an implementation of CommitmentData that treats the commitment as an opaque bytestring.
```

File: op-node/rollup/derive/plasma_data_source.go [hunk 4]
```diff
@@ -61,11 +61,18 @@ func (s *PlasmaDataSource) Next(ctx context.Context) (eth.Data, error) {
 		}
 
 		// validate batcher inbox data is a commitment.
-		comm, err := plasma.DecodeKeccak256(data[1:])
+		// strip the version byte from the data before decoding.
```
Comment (trianglesphere) [inline op-node/rollup/derive/plasma_data_source.go]:
```suggestion
		// strip the transaction data version byte from the data before decoding.
```

File: op-node/rollup/derive/plasma_data_source.go [hunk 5]
```diff
@@ -61,11 +61,18 @@ func (s *PlasmaDataSource) Next(ctx context.Context) (eth.Data, error) {
...
+		comm, err := plasma.DecodeCommitmentData(data[1:])
 		if err != nil {
 			s.log.Warn("invalid commitment", "commitment", data, "err", err)
 			return s.Next(ctx)
 		}
+		// only support keccak256 commitments for now.
+		// TODO: support other commitment types via flag
+		if comm.CommitmentType() != plasma.Keccak256CommitmentType {
+			s.log.Warn("wrong commitment type", "commitmentType", comm.CommitmentType())
+			return s.Next(ctx)
```
Comment (trianglesphere) [inline op-node/rollup/derive/plasma_data_source.go]:
```suggestion
			return nil, NotEnoughData
```
My bad for not catching this in the other error check. We should return not enough data rather than looping here.

File: op-node/rollup/derive/plasma_data_source.go [hunk 6]
```diff
@@ -61,11 +61,18 @@ func (s *PlasmaDataSource) Next(ctx context.Context) (eth.Data, error) {
 		}
 
 		// validate batcher inbox data is a commitment.
-		comm, err := plasma.DecodeKeccak256(data[1:])
+		// strip the version byte from the data before decoding.
+		comm, err := plasma.DecodeCommitmentData(data[1:])
 		if err != nil {
 			s.log.Warn("invalid commitment", "commitment", data, "err", err)
 			return s.Next(ctx)
```
Comment (trianglesphere) [inline op-node/rollup/derive/plasma_data_source.go]: scope creep: This should be changed as below

File: op-node/rollup/derive/plasma_data_source_test.go [hunk 7]
```diff
@@ -123,8 +124,11 @@ func TestPlasmaDataSource(t *testing.T) {
 			// mock input commitments in l1 transactions
 			input := testutils.RandomData(rng, 2000)
 			comm, _ := storage.SetInput(ctx, input)
+			fmt.Println("comm", comm.CommitmentType())
```
Comment (trianglesphere) [inline op-node/rollup/derive/plasma_data_source_test.go]: nit: extra println

File: op-plasma/commitment.go (hunk 3 above)
Comment (trianglesphere) [reply op-plasma/commitment.go]: Doc string still needs to be updated

File: op-plasma/commitment.go [hunk 8]
```diff
@@ -16,12 +17,85 @@ var ErrCommitmentMismatch = errors.New("commitment mismatch")
...
+}
+
+// DecodeCommitmentData parses the commitment into a known commitment type.
+// The input type is determined by the first byte of the raw data.
+// The input type is discarded and the commitment is passed to the appropriate constructor.
+func DecodeCommitmentData(input []byte) (CommitmentData, error) {
+	if len(input) == 0 {
+		return nil, ErrInvalidCommitment
+	}
+	fmt.Println(input, input[0])
```
Comment (trianglesphere) [inline op-plasma/commitment.go]: println

File: op-plasma/commitment.go [hunk 9]
```diff
@@ -16,12 +17,85 @@ var ErrCommitmentMismatch = errors.New("commitment mismatch")
...
+	t := CommitmentType(input[0])
+	data := input[1:]
+	switch t {
+	case Keccak256CommitmentType:
+		fmt.Println("DecodeKeccak256")
```
Comment (trianglesphere) [inline op-plasma/commitment.go]: ditto println

File: op-plasma/commitment.go [hunk 10]
```diff
@@ -16,12 +17,85 @@ var ErrCommitmentMismatch = errors.New("commitment mismatch")
...
+		return DecodeKeccak256(data)
+	case GenericCommitmentType:
+		fmt.Println("DecodeGenericCommitment")
```
Comment (trianglesphere) [inline op-plasma/commitment.go]: ditto

File: op-plasma/commitment.go [hunk 11]
```diff
@@ -40,22 +114,34 @@ func (c Keccak256Commitment) Verify(input []byte) error {
...
+func (c GenericCommitment) Encode() []byte {
+	return append([]byte{byte(GenericCommitmentType)}, c...)
+}
+
+// TxData adds an extra version byte to signal it's a commitment.
+func (c GenericCommitment) TxData() []byte {
+	return append([]byte{TxDataVersion1}, c.Encode()...)
+}
+
+func (c GenericCommitment) Verify(input []byte) error {
```
Comment (trianglesphere) [inline op-plasma/commitment.go:142]:
```suggestion
// Verify always returns true for GenericCommitment because the DA Server must validate the data before returning it to the op-node.
func (c GenericCommitment) Verify(input []byte) error {
```

File: op-plasma/daclient.go [hunk 12]
```diff
@@ -60,11 +60,12 @@ func (c *DAClient) GetInput(ctx context.Context, comm Keccak256Commitment) ([]by
 }
 
 // SetInput sets the input data and returns the keccak256 hash commitment.
-func (c *DAClient) SetInput(ctx context.Context, img []byte) (Keccak256Commitment, error) {
+func (c *DAClient) SetInput(ctx context.Context, img []byte) (CommitmentData, error) {
 	if len(img) == 0 {
 		return nil, ErrInvalidInput
 	}
-	comm := Keccak256(img)
+	// TODO: this is hard-coded to produce Keccak256 commitments
```
Comment (trianglesphere) [inline op-plasma/daclient.go]:
```suggestion
	// TODO(#10312): this is hard-coded to produce Keccak256 commitments
```

**Review by trianglesphere — COMMENTED** [review]
The design looks really good. I just found some small things

**Review by trianglesphere — APPROVED** [review]
